ui/app/contexts/metametrics.new.js

```diff
@@ -0,0 +1,220 @@
+ * metrics system. This file implements Segment analytics tracking.
+ */
+import React, { useRef, Component, createContext, useEffect, useCallback } from 'react'
+import { useSelector } from 'react-redux'
+import PropTypes from 'prop-types'
+import { useLocation, matchPath, useRouteMatch } from 'react-router-dom'
+import { captureException, captureMessage } from '@sentry/browser'
+
+import { omit } from 'lodash'
+import {
+  getCurrentNetworkId,
+} from '../selectors/selectors'
+
+import { getEnvironmentType } from '../../../app/scripts/lib/util'
+import {
+  sendCountIsTrackable,
+  segment,
+  METAMETRICS_ANONYMOUS_ID,
+} from '../helpers/utils/metametrics.util'
+import { PATH_NAME_MAP } from '../helpers/constants/routes'
+import { getCurrentLocale } from '../ducks/metamask/metamask'
+import { txDataSelector } from '../selectors'
+
+export const MetaMetricsContext = createContext(() => {
+  captureException(
+    Error(`MetaMetrics context function was called from a react node that is not a descendant of a MetaMetrics context provider`),
+  )
+})
+
+const PATHS_TO_CHECK = Object.keys(PATH_NAME_MAP)
+
+function useSegmentContext () {
+  const match = useRouteMatch({ path: PATHS_TO_CHECK, exact: true, strict: true })
+  const locale = useSelector(getCurrentLocale)
+  const txData = useSelector(txDataSelector) || {}
+  const confirmTransactionOrigin = txData.origin
+
+  const referrer = confirmTransactionOrigin ? {
+    url: confirmTransactionOrigin,
+  } : undefined
+
+  let version = global.platform.getVersion()
+  if (process.env.METAMASK_ENVIRONMENT !== 'production') {
+    version = `${version}-${process.env.METAMASK_ENVIRONMENT}`
+  }
+
+  const page = match ? {
+    path: match.path,
+    title: PATH_NAME_MAP[match.path],
+    url: match.path,
+  } : undefined
+
+  return {
+    app: {
+      version,
+      name: 'MetaMask Extension',
+    },
+    locale,
+    page,
+    referrer,
+    userAgent: window.navigator.userAgent,
+  }
+}
+
+export function MetaMetricsProvider ({ children }) {
+  const network = useSelector(getCurrentNetworkId)
+  const metaMetricsId = useSelector((state) => state.metamask.metaMetricsId)
+  const participateInMetaMetrics = useSelector((state) => state.metamask.participateInMetaMetrics)
+  const metaMetricsSendCount = useSelector((state) => state.metamask.metaMetricsSendCount)
+  const location = useLocation()
+  const context = useSegmentContext()
+
+  // Used to prevent double tracking page calls
+  const previousMatch = useRef()
+
+  /**
+   * Anytime the location changes, track a page change with segment.
+   * Previously we would manually track changes to history and keep a
+   * reference to the previous url, but with page tracking we can see
+   * which page the user is on and their navigation path.
+   */
+  useEffect(() => {
+    const environmentType = getEnvironmentType()
+    if (location.pathname.startsWith('/initialize') || participateInMetaMetrics) {
```

Comment:
This will still send "anonymous" calls in during initialization even after explicit opt-out. This is not a _huge_ deal, but we could prevent this for checking whether `participateInMetaMetrics` is `false` or completely unset (the default value is `null`).

e.g.:

```suggestion
    if (
      (participateInMetaMetrics === null && location.pathname.startsWith('/initialize'))
      || participateInMetaMetrics
    ) {
```